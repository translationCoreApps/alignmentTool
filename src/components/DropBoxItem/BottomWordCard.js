import React, {Component} from 'react';
import PropTypes from 'prop-types';
import { DragSource } from 'react-dnd';
import ItemTypes from '../ItemTypes';

const internalStyle = {
  display: 'flex',
  width: '100%',
  borderLeft: '5px solid #44C6FF',
  padding: '10px',
  textAlign: 'center',
  backgroundColor: '#FFFFFF',
  boxShadow: "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
  cursor: 'move'
};

class BottomWordCard extends Component {
  render() {
    const { word, style, isDragging, connectDragSource } = this.props;
    const opacity = isDragging ? 0.4 : 1;

    return connectDragSource(
      <span style={{ ...internalStyle, ...style, opacity }}>
        {word}
      </span>
    );
  }
}

BottomWordCard.propTypes = {
  word: PropTypes.string.isRequired,
  occurrence: PropTypes.number.isRequired,
  occurrences: PropTypes.number.isRequired,
  alignmentIndex: PropTypes.number,
  connectDragSource: PropTypes.func.isRequired,
  isDragging: PropTypes.bool.isRequired,
  style: PropTypes.object
};

const DragBottomWordCardAction = {
  beginDrag(props) {
    // Return the data describing the dragged item
    const item = {
      word: props.word,
      occurrence: props.occurrence,
      occurrences: props.occurrences,
      alignmentIndex: props.alignmentIndex
    };
    return item;
  }
};

const collect = (connect, monitor) => {
  return {
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(),
    isDragging: monitor.isDragging()
  };
};

export default DragSource(
  ItemTypes.BOTTOM_WORD,
  DragBottomWordCardAction,
  collect
)(BottomWordCard);
