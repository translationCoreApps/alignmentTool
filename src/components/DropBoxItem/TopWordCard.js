import React from 'react';
import PropTypes from 'prop-types';

const internalStyle = {
  borderLeft: '5px solid #44C6FF',
  padding: '10px',
  marginBottom: '5px',
  color: '#ffffff',
  textAlign: 'center',
  backgroundColor: '#333333',
  boxShadow: "0 1px 4px rgba(0, 0, 0, 0.3), 0 0 40px rgba(0, 0, 0, 0.1) inset",
  cursor: 'move'
};

const TopWordCard = ({
  words,
  style
}) => {
  return (
    <span style={{ ...internalStyle, ...style }}>
      {
        words.map((wordObject, index) => (
          <span key={index}>{wordObject.word}&nbsp;</span>
        ))
      }
    </span>
  );
};

TopWordCard.propTypes = {
  words: PropTypes.array.isRequired,
  style: PropTypes.object
};

export default TopWordCard;