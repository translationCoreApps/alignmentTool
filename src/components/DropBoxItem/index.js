import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DropTarget } from 'react-dnd';
// constants
import ItemTypes from '../ItemTypes';
// components
import SecondaryWord from '../Word';
import PrimaryWord from './TopWordCard';

class DropBoxItem extends Component {
  render() {
    const { alignmentIndex, canDrop, isOver, bottomWords, connectDropTarget, topWords } = this.props;
    const topStyle = {
      padding: topWords.length === 0 ? '15px 0px' : '1px 0'
    };
    const bottomStyle = {
      height: '35px',
      padding: bottomWords.length === 0 ? '15px 0px' : canDrop ? '15px 0px' :'0px',
      border: isOver && canDrop ? '3px dashed #44C6FF' : bottomWords.length === 0 ? '3px dashed #ffffff' : canDrop ? '3px dashed #ffffff' : ''
    };
    const dropBoxItemOuterStyle = {
      padding: '5px 10px',
      backgroundColor: '#DCDCDC',
      margin: '0px 10px 10px 0px',
      height: '100px'
    };
    const dropBoxItemInnerStyle = {
      display: 'flex',
      flexDirection: 'column',
      width: '230px',
      height: '70px',
      backgroundColor: '#DCDCDC'
    };
    let dropBoxDiv = (
      <div style={dropBoxItemOuterStyle}>
        <div style={dropBoxItemInnerStyle}>
          <div style={topStyle}>
            <div style={{ display: 'flex' }}>
              {
                topWords.map((wordObject, index) => (
                  <PrimaryWord
                    key={index}
                    wordObject={wordObject}
                    alignmentIndex={this.props.alignmentIndex}
                    resourcesReducer={this.props.resourcesReducer}
                    actions={this.props.actions}
                  />
                ))
              }
            </div>
          </div>
          <div style={bottomStyle}>
            {bottomWords.length > 0 &&
              <div style={{ display: 'flex' }}>
                {
                  bottomWords.map((metadata, index) => (
                    <SecondaryWord
                      key={index}
                      word={metadata.word}
                      occurrence={metadata.occurrence}
                      occurrences={metadata.occurrences}
                      alignmentIndex={alignmentIndex}
                    />
                  ))
                }
              </div>
            }
          </div>
        </div>
      </div>
    );
    const emptyDropBoxItem = topWords.length === 0 && bottomWords.length === 0;
    if (emptyDropBoxItem) {
      dropBoxItemInnerStyle.width = 10;
      dropBoxDiv = canDrop ? dropBoxDiv : <div />;
    }
    return connectDropTarget(
      dropBoxDiv
    );
  }
}

DropBoxItem.propTypes = {
  isOver: PropTypes.bool.isRequired,
  canDrop: PropTypes.bool.isRequired,
  draggingColor: PropTypes.string,
  connectDropTarget: PropTypes.func.isRequired,
  topWords: PropTypes.array.isRequired,
  bottomWords: PropTypes.array.isRequired,
  siblingTopWords: PropTypes.array,
  alignmentIndex: PropTypes.number.isRequired,
  lastDroppedItem: PropTypes.object,
  onDrop: PropTypes.func.isRequired,
  resourcesReducer: PropTypes.shape({
    lexicons: PropTypes.object.isRequired
  }),
  actions: PropTypes.shape({
    showPopover: PropTypes.func.isRequired,
    loadLexiconEntry: PropTypes.func.isRequired
  })
};

const DropDropBoxItemAction = {
  canDrop(props, monitor) {
    const item = monitor.getItem();
    const alignmentEmpty = (props.topWords.length === 0 && props.bottomWords.length === 0);
    let canDrop;
    if (item.type === ItemTypes.SECONDARY_WORD) {
      const alignmentIndexDelta = props.alignmentIndex - item.alignmentIndex;
      canDrop = alignmentIndexDelta !== 0 && !alignmentEmpty;
      return canDrop;
    }
    if (item.type === ItemTypes.PRIMARY_WORD) {
      const alignmentIndexDelta = props.alignmentIndex - item.alignmentIndex;
      const enoughSiblingTopWords = props.siblingTopWords && props.siblingTopWords.length > 1;
      if (alignmentIndexDelta === 0 && alignmentEmpty && enoughSiblingTopWords) {
        canDrop = true;
      } else {
        canDrop = (!alignmentEmpty && Math.abs(alignmentIndexDelta) === 1);
      }
      return canDrop;
    }
  },
  drop(props, monitor) {
    props.onDrop(monitor.getItem());
  }
};

const collect = (connect, monitor) => {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop()
  };
};

export default DropTarget(
  [ItemTypes.SECONDARY_WORD, ItemTypes.PRIMARY_WORD], // itemType
  DropDropBoxItemAction,
  collect
)(DropBoxItem);
