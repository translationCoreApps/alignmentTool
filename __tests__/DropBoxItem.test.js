import React from 'react';
import DropBoxItem from '../src/components/DropBoxItem';

test('DropBoxItem renders', () => {
    const component = (
        <DropBoxItem />
    );
    expect(component).toMatchSnapshot();

    // TODO: exercise UI
});