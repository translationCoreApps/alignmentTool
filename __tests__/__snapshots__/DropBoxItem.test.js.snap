// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropBoxItem renders 1`] = `<DropTarget(DropBoxItem) />`;
